cpp/src/cider/exec/nextgen/operators/VectorizedProjectNode.cpp

```diff
@@ -0,0 +1,262 @@
+  res.reserve(expr_groups.getUnionNum());
+  for (auto& item : groups_map) {
+    res.emplace_back(std::move(item.second));
+  }
+
+  return res;
+}
+
+jitlib::JITValuePointer& getNullBuffer(context::CodegenContext& context,
+                                       const ExprPtr& expr) {
+  CHECK(expr->getLocalIndex());
+  auto&& [_, values] = context.getArrowArrayValues(expr->getLocalIndex());
+  return utils::JITExprValueAdaptor(values).getNull();
+}
+
+jitlib::JITValuePointer& allocateNullBuffer(context::CodegenContext& context,
+                                            jitlib::JITValuePointer& len,
+                                            const ExprPtr& expr) {
+  CHECK(expr->getLocalIndex());
+  auto& arrow_array_values = context.getArrowArrayValues(expr->getLocalIndex());
+
+  utils::JITExprValueAdaptor nullable_values(arrow_array_values.second);
+  nullable_values.setNull(context.getJITFunction()->createLocalJITValue([&]() {
+    return context::codegen_utils::allocateArrowArrayBuffer(
+        arrow_array_values.first, 0, len, SQLTypes::kTINYINT);
+  }));
+
+  return nullable_values.getNull();
+}
+
+void VectorizedProjectTranslator::generateExprsGroupCode(
+    context::CodegenContext& context,
+    std::vector<ExprsGroup>& exprs_groups) {
+  for (auto& group : exprs_groups) {
+    // Utilize C2R and R2C to generate column load and store.
+    auto c2r_node = createOpNode<ColumnToRowNode>(group.input_exprs);
+    auto c2r_translator = c2r_node->toTranslator();
+    auto r2c_node = createOpNode<RowToColumnNode>(group.exprs, c2r_node.get());
+    auto r2c_translator = r2c_node->toTranslator();
+
+    {
+      // Temporarily set nullable of all exprs in output expr tree as false.
+      NullableGuard null_guard(group.exprs);
+
+      c2r_translator->codegen(
+          context, [&group, &r2c_translator](context::CodegenContext& context) {
+            // Generate Project code.
+            for (auto& expr : group.exprs) {
+              expr->codegen(context);
+            }
+            r2c_translator->consume(context);
+          });
+    }
+
+    // Process null vector.
+    // TBD (bigPYJ1151): Currently, only support trival null processing (e.g. merge all
+    // leaf exprs' null vector by bitwise-and).
+    // TODO (bigPYJ1151): Specialize more null vector 'and' primary function to reduce
+    // function-call overhead and load/store inst.
+    for (auto& expr : group.exprs) {
+      if (!expr->getNullable()) {
+        continue;
+      }
+      auto input_col = utils::collectColumnVars({expr});
+      input_col.erase(std::remove_if(input_col.begin(),
+                                     input_col.end(),
+                                     [](ExprPtr& expr) { return !expr->getNullable(); }),
+                      input_col.end());
+      CHECK(!input_col.empty());
+
+      auto output_null = allocateNullBuffer(context, c2r_node->getColumnRowNum(), expr);
+      size_t input_index = 0;
+      if (input_col.size() < 2) {
```

Comment:
maybe can mark the input null buffer as reused. and don't need allocate new buffer.

Reply:
It's fine, the buffer is allocated/reallocated through ```ArrowArrayBufferHolder```.  The ```allocateBuffer``` of ```ArrowArrayBufferHolder``` will decide allocate or reallocate.

Reply:
`allocateBuffer` reallocate according the same index, but can't reuse the input arrow array null buffer. Even if `allocatBuffer` can reuse the input buffer, memory copy can't be eliminated.

Reply:
Yes,  ```allocateBuffer``` of ```ArrowArrayBufferHolder``` will be refined to check whether need to reallocate.

Reply:
suppose the expression `a + a`, the output null equal `a`'s null, `allocateBuffer` and `memcpy` should be eliminated.